### PR TITLE
Use Webpack for building

### PR DESCRIPTION
**Pull request by stefwalter**
Start trying to use webpack. This is a work in progress with the goal to help coordinate with @larsu and help provide context for some of the related pull requests. There's a lot that doesn't work here yet, but it lets us outline what we need to do.

Goal: There will be two build systems. automake/Makefile for the base and C parts of Cockpit. And webpack (plus something?) for the javascript packages.

 * [x] Figure out how to use webpack to interact with lib/
 * [x] Move base parts of Cockpit into src/
 * [x] Figure out how to use webpack to compile less and CSS files?
 * [x] Integrate webpack into automake
 * [x] Figure out how to use webpack to copy files into the distribution directory.
 * [x] Use the turing completeness of webpack.config.js and make it not so butt ugly.
 * [x] Make dist work.

If we can't use webpack as the only build process for most simple components, then we need to choose either grunt or gulp along with it.

### CODE REVIEW

**Comment by stefwalter** [comment]
Blocked on https://github.com/cockpit-project/cockpit/pull/4661

**Comment by stefwalter** [comment]
A bunch more stuff is done here. Created a loadable package.

**Comment by larsu** [comment]
Nice work, thanks!

I've ported my my own WIP branch from a few weeks back to master and pushed it at  #4667. It doesn't have the fancy Makefile.am integration that you built here.

One notable difference is, that I tried to keep the srcdir/builddir concepts out of webpack.config.js. I passed --outputpath to the binary instead. I also started using mustache from node_modules directly.

You're obviously much further than I was. Just thought I'd push it for comparison.

File: Makefile.am [hunk 1]
```diff
@@ -322,7 +348,6 @@ include pkg/playground/Makefile.am
 include pkg/realmd/Makefile.am
 include pkg/selinux/Makefile.am
 include pkg/shell/Makefile.am
-include pkg/sosreport/Makefile.am
```
Comment (larsu) [inline Makefile.am]: This means `pkg/sosreport/Makefile.am` can be removed, right?

File: pkg/sosreport/index.js [hunk 2]
```diff
@@ -17,18 +17,21 @@
  * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
  */
 
-require([
-    "jquery",
-    "base1/cockpit",
-    "data!sosreport/run-sosreport.sh",
-    "shell/po"
-], function($, cockpit, run_sosreport_sh, po) {
+(function() {
```
Comment (larsu) [inline pkg/sosreport/index.js]: Do we still need the function wrapper when using webpack?

File: pkg/sosreport/override.json [hunk 3]
```diff
@@ -0,0 +1,3 @@
+{
+	"base": "../../dist/pkg/sosreport"
+}
```
Comment (larsu) [inline pkg/sosreport/override.json]: Hm, what does this configure?

File: pkg/sosreport/index.js (hunk 2 above)
Comment (stefwalter) [inline pkg/sosreport/index.js]: Yes, so that module specific globals don't leak into the global scope.

File: Makefile.am (hunk 1 above)
Comment (stefwalter) [inline Makefile.am]: Right. Removed.

**Comment by stefwalter** [comment]
distcheck doesn't pass yet: https://travis-ci.org/cockpit-project/cockpit/jobs/142493580

**Comment by dperpeet** [comment]
still depends on #4661 
Please clarify if that one should just be closed and merged in one go here?

**Comment by stefwalter** [comment]
>  Please clarify if that one should just be closed and merged in one go here?

One at a time is fine with me.

**Comment by mvollmer** [comment]
#4661 is merged

File: pkg/sosreport/index.js (hunk 2 above)
Comment (larsu) [inline pkg/sosreport/index.js]: Webpack puts every module in its own function anyway, doesn't it?

**Comment by stefwalter** [comment]
I'm assuming I can add more fixes here. Tweaks made while migrating other modules.

**Comment by stefwalter** [comment]
Temporarily lost some changes due to a rebase. Brought them back in.

**Comment by larsu** [comment]
I had trouble building this branch out of tree, because webpack can't find its loaders. I have a commitfixing this in [larsu/cockpit/webpack-out-of-tree](/larsu/cockpit/commits/webpack-out-of-tree).

(And another one fixing a typo.)

**Comment by stefwalter** [comment]
Squashed the typo pull request. And adjusted the resolveLoader change so that it works with in-tree build subdirectory.

**Comment by stefwalter** [comment]
Added one more fix.

**Comment by stefwalter** [comment]
Blocked on https://github.com/cockpit-project/cockpit/pull/4696

**Comment by stefwalter** [comment]
> Webpack puts every module in its own function anyway, doesn't it?

Perhaps it does. Certainly as we move to ES6 it works that way. I would rather take it a step at a time though, and leave the scoping for now. Especially as many modules are shared between webpack and non-webpack bundling.

**Pull request closed without merging**